torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -2305,8 +2305,78 @@ class AsyncAlltoallWork : public ProcessGroupGloo::AsyncWork {
       gloo::alltoallv(opts);
     }
   }
+
+  void run() override {
+    alltoall(outputTensor, inputTensor);
+  }
+};
+
+#ifdef USE_CUDA
+
+class AsyncAlltoallCUDAWork : public AsyncAlltoallWork {
+ public:
+  AsyncAlltoallCUDAWork(
+      const std::shared_ptr<gloo::Context>& context,
+      at::Tensor& outputTensor,
+      at::Tensor& inputTensor,
+      std::vector<int64_t>& outputCounts,
+      std::vector<int64_t>& inputCounts,
+      uint32_t tag)
+      : AsyncAlltoallWork(
+            context,
+            outputTensor,
+            inputTensor,
+            outputCounts,
+            inputCounts,
+            tag) {
+    initializeStreamsEvents({inputTensor}, inputStreams, inputEvents);
+    initializeStreamsEvents({outputTensor}, outputStreams, outputEvents);
+
+    // Kick off copy from CUDA tensors to pinned CPU tensors.
+    at::cuda::OptionalCUDAStreamGuard guard;
+    guard.reset_stream(inputStreams.front());
+    tmpInput = pinnedLike(inputTensor).copy_(inputTensor, true);
+
+    guard.reset_stream(outputStreams.front());
+    tmpOutput = pinnedLike(outputTensor);
+  }
+
+  void run() override {
+    // Synchronize with copy operations.
+    at::cuda::OptionalCUDAGuard device_guard;
+    device_guard.set_index(inputTensor.get_device());
+    AT_CUDA_CHECK(cudaStreamSynchronize(inputStreams.front()));
+    device_guard.set_index(outputTensor.get_device());
+    AT_CUDA_CHECK(cudaStreamSynchronize(outputStreams.front()));
+
+    // Run alltoall on host side tensors.
+    alltoall(tmpOutput, tmpInput);
+
+    // Kick off copy back to the CUDA tensors.
+    at::cuda::OptionalCUDAStreamGuard stream_guard;
+    stream_guard.reset_stream(outputStreams.front());
+    outputTensor.copy_(tmpOutput, /* non_blocking */ true);
+    outputEvents.front().record(outputStreams.front());
+  }
+
+  void synchronize() override {
+    // Synchronize with the copy back to CUDA tensors.
+    at::cuda::OptionalCUDAGuard guard;
+    guard.set_index(static_cast<at::DeviceIndex>(outputTensor.get_device()));
+    outputEvents.front().block(at::cuda::getCurrentCUDAStream());
+  }
+
+  at::Tensor tmpOutput;
```

Comment:
nit: would these be better named `cpuInput`/`cpuOutput` to reflect that they are CPU tensors?